### PR TITLE
fix: remove injectIntl types from component props

### LINKED ISSUE #1414: 3.1.7 break compatibility
```
TS2741: Property 'intl' is missing in type .....
```

Need to remove the use of `WrappedComponentProps`.

WrappedComponentProps definition:
https://github.com/formatjs/react-intl/blob/79647dae125c5af15ce58afd1899cd6a4c13cd29/src/components/injectIntl.tsx#L28-L30

Affected code:
https://github.com/formatjs/react-intl/blob/ce560e73b159ffd332fa2ded989f12d758b6bb1b/src/components/message.tsx#L43-L45
https://github.com/formatjs/react-intl/blob/ce560e73b159ffd332fa2ded989f12d758b6bb1b/src/components/relative.tsx#L63-L65

### PR DESCRIPTION
**Pull request by emmenko**
Leftover of #1413 

Fixes #1414

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -16,15 +16,9 @@ export {
...
-export const {Component: FormattedDate} = createFormattedComponent(
-  'formatDate'
-);
-export const {Component: FormattedTime} = createFormattedComponent(
-  'formatTime'
-);
-export const {Component: FormattedNumber} = createFormattedComponent(
-  'formatNumber'
-);
+export const FormattedDate = createFormattedComponent('formatDate');
```
Comment (septs) [inline src/index.ts]:
```
interface Formatter {
    formatDate: FormatDateOptions;
    formatTime: FormatDateOptions;
    formatNumber: FormatNumberOptions;
}

function createFomattedComponent<K extends keyof Formatter>(name: K) {
   type Options = Formatter[K]; // get options typing
}
```

Comment (emmenko) [reply src/index.ts]: Good idea, thanks!

Comment (emmenko) [reply src/index.ts]: https://github.com/formatjs/react-intl/pull/1415/commits/f0f4b7854b157418d4808b95f192889e00544168

File: src/components/createFormattedComponent.tsx [hunk 2]
```diff
@@ -3,13 +3,6 @@ import {invariantIntlContext} from '../utils';
...
 import {Context} from './injectIntl';
 
-type FormatFunctionName = 'formatDate' | 'formatTime' | 'formatNumber';
-type FormatOptions<T> = T extends 'formatDate'
-  ? FormatDateOptions
-  : T extends 'formatTime'
-  ? FormatDateOptions
-  : FormatNumberOptions;
-
 const displayNames: {
```
Comment (septs) [inline src/components/createFormattedComponent.tsx]:
```
enum DisplayName {
   formatDate = 'FormattedDate';
   formatTime = 'FormattedTime';
   formatNumber = 'FormattedNumber';
}

DisplayName['formatDate']// the value is "FormattedDate"
```

File: src/components/createFormattedComponent.tsx [hunk 3]
```diff
@@ -20,11 +13,18 @@ const displayNames: {
...
+  formatTime: FormatDateOptions;
+  formatNumber: FormatNumberOptions;
+};
+
+export default function createFomattedComponent<Name extends keyof Formatter>(
+  name: Name
 ) {
-  type FormatFn = IntlShape[T];
-  type Props = FormatOptions<T> & {
+  type Options = Formatter[Name]; // this is options typing
```
Comment (septs) [inline src/components/createFormattedComponent.tsx]: please remove comments.

File: src/components/createFormattedComponent.tsx (hunk 2 above)
Comment (emmenko) [reply src/components/createFormattedComponent.tsx]: I would personally avoid using enums

File: src/components/createFormattedComponent.tsx [hunk 4]
```diff
@@ -1,46 +1,54 @@
...
   };
+
+  const Component = (props: Props) => (
+    <Context.Consumer>
+      {intl => {
+        invariantIntlContext(intl);
+
+        const {value, children} = props;
+        const {[name]: formatFn, textComponent: Text} = intl;
+        const formattedValue = formatFn(value as any, props);
```
Comment (septs) [inline src/components/createFormattedComponent.tsx]: why not `intl[name](value as any, props)`?

File: src/components/createFormattedComponent.tsx (hunk 2 above)
Comment (septs) [reply src/components/createFormattedComponent.tsx]: why?

File: src/components/createFormattedComponent.tsx (hunk 4 above)
Comment (emmenko) [reply src/components/createFormattedComponent.tsx]: I left this logic as before to be honest, but yeah I can change that as well.

File: src/components/createFormattedComponent.tsx (hunk 2 above)
Comment (septs) [reply src/components/createFormattedComponent.tsx]: please use https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions

Comment (septs) [reply src/components/createFormattedComponent.tsx]: image

Comment (emmenko) [reply src/components/createFormattedComponent.tsx]: Thanks for the hint (I didn't know about this feature). However, this project is currently using TS 3.3.

I plan to open an issue to list the tooling and setup improvements (e.g. latest TS, using RTL for tests, etc).

Comment (septs) [reply src/components/createFormattedComponent.tsx]: Got

File: src/components/createFormattedComponent.tsx (hunk 4 above)
Comment (septs) [reply src/components/createFormattedComponent.tsx]: OK

File: src/components/createFormattedComponent.tsx [hunk 5]
```diff
@@ -1,46 +1,52 @@
...
+    <Context.Consumer>
+      {intl => {
+        invariantIntlContext(intl);
+
+        const formattedValue = intl[name](props.value as any, props);
+
+        if (typeof props.children === 'function') {
+          return props.children(formattedValue);
+        }
+        if (intl.textComponent) {
```
Comment (septs) [inline src/components/createFormattedComponent.tsx]:
```typescript
const Text = intl.textComponent || React.Fragment;
return <Text>{formattedValue}</Text>;
```
see https://reactjs.org/docs/fragments.html

Comment (emmenko) [reply src/components/createFormattedComponent.tsx]: Thanks: https://github.com/formatjs/react-intl/pull/1415/commits/669dfbbb85be1a882da3ba1a3ce7a4815479740d

File: src/components/createFormattedComponent.tsx [hunk 6]
```diff
@@ -1,46 +1,49 @@
...
+  name: Name
+) {
+  type Options = Formatter[Name];
+  type FormatFn = IntlShape[Name];
+  type Props = Options & {
+    value: Parameters<FormatFn>[0];
+    children?: (val: string) => React.ReactElement | null;
   };
+
+  const Component = (props: Props) => (
```
Comment (septs) [inline src/components/createFormattedComponent.tsx]: #1416 not use any then
```typescript
const Component: React.FC<Props> = (props) => {
   // ... code ...
}
```

`React.FC` definition: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/981cc6d51eb213943a3606c14ad75d65ea6b360f/types/react/index.d.ts#L493-L501

Comment (emmenko) [reply src/components/createFormattedComponent.tsx]: Hmm what issue should this solve exactly? Also, can you explain how using `React.FC` is different from `props: Props`? 🤔

Comment (septs) [reply src/components/createFormattedComponent.tsx]: expand `React.FC<Props>` definition got `(props: Props, context?: any) => React.ReactElement | null`

Comment (longlho) [reply src/components/createFormattedComponent.tsx]: yeah `const Component: React.FC<Props> = props =>` gives u the correct type definition

File: src/components/createFormattedComponent.tsx (hunk 2 above)
Comment (septs) [reply src/components/createFormattedComponent.tsx]: @longlho use enum definition, `DisplayName`, I think it is the correct practice.

Comment (longlho) [reply src/components/createFormattedComponent.tsx]: `const enum` would work here I believe?

File: src/components/createFormattedComponent.tsx (hunk 6 above)
Comment (emmenko) [reply src/components/createFormattedComponent.tsx]: Just to clarify, what do you mean by `the correct type definition`?
Using `props: Props` is not wrong, this is also what you [find in the TS docs](https://www.typescriptlang.org/docs/handbook/jsx.html#function-component)

image

I'm also not against using `React.FC` if you want to, I just want to understand what are the benefits of using that compared to `props: Props`.

Comment (emmenko) [reply src/components/createFormattedComponent.tsx]: https://github.com/formatjs/react-intl/pull/1415/commits/11335f78c3c04b91078ca311650a5dcfd1069a03

File: src/components/createFormattedComponent.tsx (hunk 2 above)
Comment (emmenko) [reply src/components/createFormattedComponent.tsx]: Hmm sorry, I'm not sure I understand what you need me to do here 🤔

Comment (septs) [reply src/components/createFormattedComponent.tsx]:
```typescript
const enum DisplayName {
   formatDate = 'FormattedDate',
   formatTime = 'FormattedTime',
   formatNumber = 'FormattedNumber',
}
```

in component:
```
Component.displayName = DisplayName[name];
```

Comment (emmenko) [reply src/components/createFormattedComponent.tsx]: I'm getting this error: 

image

Comment (septs) [reply src/components/createFormattedComponent.tsx]: replace `const enum` to `enum`

Comment (emmenko) [reply src/components/createFormattedComponent.tsx]: Ah thanks, I never used `enum` so my knowledge of it is a bit limited.

https://github.com/formatjs/react-intl/pull/1415/commits/41d319fa1fe8ff415810645078f834e779d083f8

**Pull request merged**